Sources/NIOEmbedded/AsyncEmbeddedEventLoop.swift

```diff
@@ -0,0 +1,405 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+import Dispatch
+import _NIODataStructures
+import NIOCore
+import NIOConcurrencyHelpers
+
+
+/// An `EventLoop` that is thread safe and whose execution is fully controlled
+/// by the user.
+///
+/// Unlike more complex `EventLoop`s, such as `SelectableEventLoop`, the `NIOAsyncEmbeddedEventLoop`
+/// has no proper eventing mechanism. Instead, reads and writes are fully controlled by the
+/// entity that instantiates the `NIOAsyncEmbeddedEventLoop`. This property makes `NIOAsyncEmbeddedEventLoop`
+/// of limited use for many application purposes, but highly valuable for testing and other
+/// kinds of mocking. Unlike `EmbeddedEventLoop`, `NIOAsyncEmbeddedEventLoop` is fully thread-safe and
+/// safe to use from within a Swift concurrency context.
+///
+/// Unlike `EmbeddedEventLoop`, `NIOAsyncEmbeddedEventLoop` does require that user tests appropriately
+/// enforce thread safety. Used carefully it is possible to safely operate the event loop without
+/// explicit synchronization, but it is recommended to use `executeInContext` in any case where it's
+/// necessary to ensure that the event loop is not making progress.
+///
+/// Time is controllable on an `NIOAsyncEmbeddedEventLoop`. It begins at `NIODeadline.uptimeNanoseconds(0)`
+/// and may be advanced by a fixed amount by using `advanceTime(by:)`, or advanced to a point in
+/// time with `advanceTime(to:)`.
+///
+/// If users wish to perform multiple tasks at once on an `NIOAsyncEmbeddedEventLoop`, it is recommended that they
+/// use `executeInContext` to perform the operations. For example:
+///
+/// ```
+/// await loop.executeInContext {
+///     // All three of these will be queued up simultaneously, and no other code can
+///     // get between them.
+///     loop.execute { firstTask() }
+///     loop.execute { secondTask() }
+///     loop.execute { thirdTask() }
+/// }
+/// ```
+///
+/// There is a tricky requirement around waiting for `EventLoopFuture`s when working with this
+/// event loop. Simply calling `.wait()` from the test thread will never complete. This is because
+/// `wait` calls `loop.execute` under the hood, and that callback cannot execute without calling
+/// `loop.run()`. As a result, if you need to await an `EventLoopFuture` created on this loop you
+/// should use `awaitFuture`.
```

Comment:
Do we want to add a callback to the `EventLoop` protocol, to check if an EventLoop supports awaiting a future without explicitly running? If it doesn't support it, we could crash in `wait()`.

Reply:
As discussed on the call I think the answer is "no": for now I don't think the diagnostic information we gather is sufficiently useful. We can always add this later if it turns out to be a problem.